### ISSUE TITLE
Reentrancy rules was removed?

### ISSUE THREAD

**Issue by ggrieco-tob** [smartdec/smartcheck#5]
I noticed that the reentrancy rule (SOLIDITY_REENTRANCY_EXTERNAL_CALL) was [removed](https://github.com/smartdec/smartcheck/commit/30871c70b4cebc5450ee614029300c0635f57dfb#diff-83e8fc57529e51e34c5680ab50bc3778). Is there any plans to add it back?

**Comment by eMarchenko** [comment]
Yes, it was removed, because it was producing too many false positives.
We will check if we can improve that rule enough to restore it.
Even if we don't like the result we can share our best attempt here or as commented out rule.

**Comment by ggrieco-tob** [comment]
Fair enough. Feel free to close this issue if there is no plans to re-add it.

**Comment by eMarchenko** [comment]
By the way, reentrancy is more like bunch of similar issues, though the most well-known one is "the DAO"-style. We are currently working on several cases, but writing proper xpath for reentrancy proves to be difficult. 
Could you provide an example of exact reentrancy type you are interested in? So we include it in our patterns as well.

**Comment by eMarchenko** [comment]
@ggrieco-tob  you can try this rule
```
<Rule>
    <RuleId>SOLIDITY_REENTRANCY_EXTERNAL_CALL</RuleId>
    <Patterns>
        <!-- Looks for functions with side effect (sstrore or delegatecall) after external call. -->
        <!-- Фиксируется случаи, когда внутри определения функции после внешнего вызова(кроме SafeMath и функций с названиями pop, push)
        есть изменение state переменных(с операторами '|=' , '^=' , '&=' , '+=' , '-=' , '/=' , '*=' , '%=', '<<=' , '>>='). -->
        <Pattern patternId="de222d">
            <Categories>
                <Category>Solidity</Category>
            </Categories>
            <Severity>3</Severity>
            <XPath>
                //functionDefinition[block/statement/simpleStatement//expression <!-- function -->
                    [lvalueOperator or text()[1]="=" or (text()[1] = "." and functionCall/functionName/identifier[text()[1]="delegatecall"])] <!-- that has side effect (assignment or delegatecall) ... -->
                    [ancestor::statement[preceding-sibling::*[simpleStatement//expression <!-- ... after external call: -->
                        [functionCall/functionName/identifier[not(text()[1] = "push") and not(text()[1] = "pop") and not(text()[1] = "add") and not(text()[1] = "sub") and not(text()[1] = "mul") and not(text()[1] = "div")]] <!-- not SafeMath and not push/pop -->
                        [text()[1] = "."] <!--< identifier>.<call()> -->
                        [expression//identifier[not(text()[1] = "this") and not(text()[1] = "super")]] <!-- eliminate this.foo() and super.foo() cases -->
                    ]]]
                    [expression//identifier[ <!-- and the potential side effect ... -->
                        not(ancestor::block[statement//variableDeclaration] and text()[1] = ancestor::block//variableDeclaration/identifier[text()[1]]) <!-- ... is not an assignment to local variable ... -->
                        and not(ancestor::functionDefinition/parameterList[parameter] and text()[1] = ancestor::functionDefinition/parameterList//identifier[text()[1]]) <!-- ... and is not an assignment to function's parameter ... -->
                        and not(ancestor::functionDefinition[returnsParameters] and ancestor::functionDefinition/returnsParameters//identifier[text()[1]] = text()[1]) <!-- ... and is not an assignment to function's return parameter -->
                    ]]
                ]
            </XPath>
        </Pattern>
    </Patterns>
</Rule>
```
just add it to `src/main/resources/solidity-rules.xml`

**Comment by xf97** [comment]
Hello, I am a student who wants to use smartcheck. I made a mistake when installing SmartCheck using npm install @smartdec / smartcheck -g. The system environment is win10. Could I  install SmartCheck in this way now?
